sdk/core/azure-core/src/base64.cpp

```diff
@@ -4,93 +4,474 @@
+    destinationIndex += 3;
+    sourceIndex += 4;
+  }
+
+  // We are guaranteed to have an input with at least 4 bytes at this point, with a size that is a
+  // multiple of 4.
+  int64_t i0 = *(inputPtr + inputSize - 4);
+  int64_t i1 = *(inputPtr + inputSize - 3);
+  int64_t i2 = *(inputPtr + inputSize - 2);
+  int64_t i3 = *(inputPtr + inputSize - 1);
+
+  i0 = Base64DecodeArray[i0];
+  i1 = Base64DecodeArray[i1];
+
+  i0 <<= 18;
+  i1 <<= 12;
+
+  i0 |= i1;
+
+  if (i3 != EncodingPath)
   {
-    std::vector<uint8_t> decoded;
-    // According to RFC 4648, the encoded length should be ceiling(n / 3) * 4, so we can infer an
-    // upper bound here
-    DWORD decodedLength = DWORD(text.length() / 4 * 3);
-    decoded.resize(decodedLength);
-
-    CryptStringToBinaryA(
-        text.data(),
-        static_cast<DWORD>(text.length()),
-        CRYPT_STRING_BASE64 | CRYPT_STRING_STRICT,
-        reinterpret_cast<BYTE*>(decoded.data()),
-        &decodedLength,
-        nullptr,
-        nullptr);
-    decoded.resize(decodedLength);
-    return decoded;
+    i2 = Base64DecodeArray[i2];
+    i3 = Base64DecodeArray[i3];
+
+    i2 <<= 6;
+
+    i0 |= i3;
+    i0 |= i2;
+
+    base64WriteThreeLowOrderBytes(destinationPtr, i0);
+    destinationPtr += 3;
   }
+  else if (i2 != EncodingPath)
+  {
+    i2 = Base64DecodeArray[i2];
 
-#elif defined(AZ_PLATFORM_POSIX)
+    i2 <<= 6;
 
-  std::string Convert::Base64Encode(const std::vector<uint8_t>& data)
+    i0 |= i2;
+
+    *(destinationPtr + 1) = static_cast<uint8_t>(i0 >> 8);
+    *destinationPtr = static_cast<uint8_t>(i0 >> 16);
+    destinationPtr += 2;
+  }
+  else
   {
-    BIO* bio = BIO_new(BIO_s_mem());
-    bio = BIO_push(BIO_new(BIO_f_base64()), bio);
-    BIO_set_flags(bio, BIO_FLAGS_BASE64_NO_NL);
-    BIO_write(bio, data.data(), static_cast<int>(data.size()));
-    (void)BIO_flush(bio);
-    BUF_MEM* bufferPtr;
-    BIO_get_mem_ptr(bio, &bufferPtr);
-    std::string toReturn(bufferPtr->data, bufferPtr->length);
-    BIO_free_all(bio);
-
-    return toReturn;
+    *destinationPtr = static_cast<uint8_t>(i0 >> 16);
+    destinationPtr += 1;
   }
 
-  std::vector<uint8_t> Convert::Base64Decode(const std::string& text)
+  auto resultSize = static_cast<size_t>(destinationPtr - destination.begin());
+  if (resultSize < destination.size())
   {
-    std::vector<uint8_t> decoded;
-    // According to RFC 4648, the encoded length should be ceiling(n / 3) * 4, so we can infer an
-    // upper bound here
-    size_t maxDecodedLength = text.length() / 4 * 3;
-    decoded.resize(maxDecodedLength);
-
-    BIO* bio = BIO_new_mem_buf(text.data(), -1);
-    bio = BIO_push(BIO_new(BIO_f_base64()), bio);
-    BIO_set_flags(bio, BIO_FLAGS_BASE64_NO_NL);
-    int decodedLength = BIO_read(bio, &decoded[0], static_cast<int>(text.length()));
-    BIO_free_all(bio);
-
-    decoded.resize(decodedLength);
-    return decoded;
+    destination.resize(resultSize);
+    destination.shrink_to_fit();
```

Comment:
Resizing strings and vectors after the fact seems to be a relatively expensive operation.
@CaseyCarter how do we avoid this, and maybe just create the vector of the right size at the end without having to do linear data copy or re-allocation? In .NET/C#, this was relatively easy with span/slicing for buffers.